### DISCUSSION TITLE
Segmentation fault for ccache when building AOSP

### DISCUSSION THREAD

**Discussion by rhoeasy** [ccache/ccache#1102] [Support]
I used **ccache **v4.6.1**** in building **AOSP**. It reported the **segemntation fault**, then I restart the building right away with ccache. The result is successfully. Code is the same as last time. What I could do to debug？
Is export **CCACHE_DEBUG**, **CCACHE_DEBUGDIR** these two environment variables helpful? The manual said these could help explained why don't get missed, and I thought this problem may be a bug of ccache itself.
[segmentation log.txt](https://github.com/ccache/ccache/files/8976893/segmentation.log.txt)

**Comment by jrosdahl** [comment]
CCACHE_DEBUG and CCACHE_DEBUGDIR are useful when debugging cache misses, not to debug crashes. To debug a ccache segfault you would need a ccache executable built with debug symbols. Then you can collect the resulting core file produced by the OS when ccache crashed and run e.g. GDB to see where it crashed.

That said, it kind of looks like the log you attached is from a Bazel sandbox. Is that correct? If so, Bazel does it's own caching and you would likely be better off not using ccache at all if you're building with Bazel.

**Reply by rhoeasy** [reply]
Thank you for your advice very much, it may be a little difficult for me to debug. Choosing other versions of ccache may be a good solution for me, I thought.

By the way, I think the log should be outputs of ninja rather than Bazel, and ccache is useful for reducing the building time. Thank you again.

**Comment by rhoeasy** [comment]
Hi, I had the same problem again. I compiled a ccache executable whose version name is RelWithDebInfo, is it correct? But I don't know how to make it generate a core dump file. I don't found any information about how to use this version.
`$dmesg -T`
![image](https://user-images.githubusercontent.com/15611302/181216279-d05fdb83-47c0-415a-aba2-846fa342be14.png)
the picture above is my dmesg information. 

I found that ccache crashes when calling `libc-2.27.so`. Let the instruction pointer's value is `0x00007fda1eee947a`, the base address of `libc-2.27.so` is `0x7fda1ee6a000`, then subtract the base address of the libc-2.27.so library from the value of the instruction pointer to get the address where the code error occurred.

`0x7fda1eee947a - 0x7fda1ee6a000 = 0x7f47a`

nextly, I run the command `$objdump -d libc-2.27.so`. And I thought the address of `0x7f47a` could be the bug's root cause. 
![image](https://user-images.githubusercontent.com/15611302/181219410-64caf055-614f-4a1b-8d13-cdfc9e04d7c5.png)

I could provide more information about this bug, could you give me some help with debugging this problem?

1. ccache version: 4.6.1 RelWithDebInfo.
2. gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)
3. operating system: Linux  5.4.54-2 #2 SMP Thu Sep 24 17:18:11 CST 2020 x86_64 x86_64 x86_64 GNU/Linux

**Reply by jrosdahl** [reply]
> I compiled a ccache executable whose version name is RelWithDebInfo, is it correct?

That sounds strange. The version name of a custom build should either be just the version (4.6.1 in this case) or a branch + a hash (e.g. master.17dd96b4). But this shouldn't matter for your problem.

That said, to make debugging easier (when you are able to produce a core dump filei) you could build with `-D CMAKE_BUILD_TYPE=Debug` instead of `-D CMAKE_BUILD_TYPE=RelWithDebInfo`.

> But I don't know how to make it generate a core dump file.

Perhaps see some guide in the net, like https://medium.com/@sourabhedake/core-dumps-how-to-enable-them-73856a437711?

> I could provide more information about this bug, could you give me some help with debugging this problem?

I'd say: either learn how to enable core dumps and debug ccache with that core dump, or provide detailed information about how to  reproduce the problem so that somebody else can trigger the bug and debug it.

If you provide which commands I can use to check out and build the code and then trigger the problem then I can have a look.

**Reply by rhoeasy** [reply]
Thank you for your support and help. Since this problem doesn't occur every time, I don't think it's helpful to provide commands to reproduce the bug. I'll debug by trying to generate a core dump file and then post it.

**Reply by jrosdahl** [reply]
> Since this problem doesn't occur every time, I don't think it's helpful to provide commands to reproduce the bug. 

Why not?

> I'll debug by trying to generate a core dump file and then post it.

Such a core dump will most likely only be usable by yourself, so there is no need to post it. What you _can_ do is to load the code dump in a debugger and print a stack trace – that is potentially useful for others.

**Comment by rhoeasy** [comment]
Hi, I built a ccache executable with `-D CMAKE_BUILD_TYPE=Debug` and got the core file.

The picture below is the information I debugged with the command
`$gdb ./ccache core`
![image](https://user-images.githubusercontent.com/15611302/184132905-71bb4b31-5d32-43e3-a64b-acd94796983d.png)

If there is some other information need to supply I'm pleased to offer it. 

Have a nice day!

**Reply by jrosdahl** [reply]
Thanks. There is clearly a race condition here: The call at https://github.com/ccache/ccache/blob/42b1376142efc2e8ce647b660a27093d934be3de/src/ccache.cpp#L1872-L1873 has found a result file (a file in the cache), but the result file can be removed before reaching https://github.com/ccache/ccache/blob/42b1376142efc2e8ce647b660a27093d934be3de/src/ccache.cpp#L1879 which then fails to open `*result_path` and`file.get()` on https://github.com/ccache/ccache/blob/42b1376142efc2e8ce647b660a27093d934be3de/src/ccache.cpp#L1880 will return `nullptr` (this is the value of `m_stream` in stack frame #&#8203;1), leading to a crash later down in https://github.com/ccache/ccache/blob/42b1376142efc2e8ce647b660a27093d934be3de/src/ccache.cpp#L1881

The good news is that this race condition will be fixed as a side-effect of refactoring that I'm currently doing in the area. It will be included in ccache 4.7.

However, I think that it's quite strange that this happens to you more or less often since the deletion has to happen in a very small time interval and files aren't deleted from the cache that often – only on automatic cleanup or if somebody happens to delete or clear the cache by other means.

**Reply by rhoeasy** [reply]
Now I can understand why segfault happened.

I exactly set on the automatic cleanup, and the build process of AOSP must have a large number of instructions be executing at the same time, I thought.

By the way, I'm interested in if there is any information on when the ccache 4.7 will be released.

**Reply by rhoeasy** [reply]
Do you have any patches for the bug? Maybe I could help to test the race condition.

**Reply by jrosdahl** [reply]
> By the way, I'm interested in if there is any information on when the ccache 4.7 will be released.

I just wrote #1137 to answer this. :slightly_smiling_face:

> Do you have any patches for the bug? Maybe I could help to test the race condition.

As mentioned, the code will be refactored away soon, but I made a fix on master anyway: 56aac268892491465b042db51a0a3884403b92e3. Please try it out.

**Reply by jrosdahl** [reply]
After evaluating the set of fixes done on master after 4.6.1, I've decided to do a new bug fix release (4.6.2) shortly, so I'll include 56aac268892491465b042db51a0a3884403b92e3 in that.
